Update Angular CLI ivy tests

filipesilva opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
This PR updates the versions of Angular CLI and Build Optimizer used in Angular tests.

The updated versions contain improvements in code optimization that enable smaller bundle sizes with Ivy on Angular CLI and Bazel projects.


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

filipesilva committed:
build: update version of Angular CLI and Build Optimizer used for tests

cexbrayat commented:
@filipesilva Dumb question: if this enables smaller bundles, shouldn't https://github.com/angular/angular/blob/master/integration/_payload-limits.json be updated with smaller numbers for `cli-hello-world-ivy`?

filipesilva commented:
@cexbrayat not a dumb question at all, but rather a very valid point I did not address. Thanks for a heads up!

filipesilva committed:
build: rename cli-hello-world-ivy test to cli-hello-world-ivy-compat

There are two ways to bootstrap an Ivy app: with `platformBrowserDynamic().bootstrapModule` and `renderComponent`.

To distinguish between these two approaches we call the `platformBrowserDynamic().bootstrapModule` way `ivy-compat` and the `renderComponent` way just `ivy`.

filipesilva committed:
build: add cli-hello-world-ivy-minimal test with renderComponent

JoostK commented on integration/_payload-limits.json:
```diff
@@ -3,7 +3,7 @@
       "uncompressed": {
         "runtime": 1497,
-        "main": 189865,
+        "main": 164671,
```
Amazing work! 🎉

gkalpak commented on integration/run_tests.sh:
```diff
@@ -61,9 +61,9 @@ for testDir in ${TEST_DIRS}; do
-    # Track payload size for cli-hello-world, cli-hello-world-ivy and hello_world__closure
-    if $CI && ([[ $testDir == cli-hello-world ]] || [[ $testDir == cli-hello-world-ivy ]] || [[ $testDir == hello_world__closure ]]); then
-      if ([[ $testDir == cli-hello-world ]] || [[ $testDir == cli-hello-world-ivy ]]); then
+    # Track payload size for cli-hello-world, cli-hello-world-ivy-compat and hello_world__closure
```
+ cli-hello-world-ivy

filipesilva replied:
Done

gkalpak reviewed (commented):
:tada:

filipesilva committed:
build: re-enable CLI integration test for prod mode e2e

filipesilva committed:
fix(compiler-cli): don't throw when listing lazy routes for an entry route

In https://github.com/angular/angular/pull/27697 the listLazyRoutes was fixed to work with ivy.

Since the entryRoute argument is not supported, it was made to also error.

But by erroring it breaks existing usage with Angular CLI where the entry route is sent in as an argument.

This commit changes listLazyRoutes to not error out, but instead ignore the argument.

mary-poppins commented:
You can preview 9d7620e at https://pr28372-9d7620e.ngbuilds.io/.

mary-poppins commented:
You can preview a943c32 at https://pr28372-a943c32.ngbuilds.io/.

gkalpak commented on integration/_payload-limits.json:
```diff
@@ -12,7 +12,16 @@
+    "master": {
+      "uncompressed": {
+        "runtime": 1440,
+        "main": 194626,
```
So, ivy-compat is larger than pre-ivy 🤔 
Is this expected?

filipesilva replied:
Just checked with @IgorMinar and it is not wholly unexpected. We already have a task to validate size and build time differences in real apps (like AIO), and we can also iterate over this size now.

gkalpak commented on integration/cli-hello-world-ivy-compat/e2e/src/app.e2e-spec.ts:
```diff
@@ -0,0 +1,19 @@
+import { AppPage } from './app.po';
+
+describe('cli-hello-world-ivy App', () => {
```
_cli-hello-world-ivy_ --> _cli-hello-world-ivy-compat_

filipesilva replied:
Fixed

mary-poppins commented:
You can preview 82f3686 at https://pr28372-82f3686.ngbuilds.io/.

gkalpak commented on integration/cli-hello-world-ivy-compat/src/index.html:
```diff
@@ -0,0 +1,14 @@
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>CliHelloWorldIvy</title>
```
_CliHelloWorldIvy_ --> _CliHelloWorldIvyCompat_

filipesilva replied:
Fixed

gkalpak commented on integration/cli-hello-world-ivy/e2e/src/app.e2e-spec.ts:
```diff
@@ -1,6 +1,10 @@
 import { AppPage } from './app.po';
 
-describe('cli-hello-world-ivy App', () => {
+describe('cli-hello-world App', () => {
```
_cli-hello-world_ --> _cli-hello-world-ivy_

filipesilva replied:
Fixed

gkalpak commented on integration/cli-hello-world-ivy/e2e/src/app.e2e-spec.ts:
```diff
@@ -1,6 +1,10 @@
 
-describe('cli-hello-world-ivy App', () => {
+describe('cli-hello-world App', () => {
+  // Ivy renderComponent apps fail on protractor when waiting for Angular.
```
Is this a known breaking change?

filipesilva replied:
Just checked with @IgorMinar, and he said it wasn't unexpected but `renderComponent` isn't how we should be doing real world Ivy apps at the moment so it wasn't worrying for now.

filipesilva replied:
Also opened an issue for it https://github.com/angular/angular/issues/28449

gkalpak commented on integration/cli-hello-world-ivy/package.json:
```diff
@@ -10,9 +10,7 @@
-    "//test1": "TODO FW-813: Re-enable `ci-production`. Currently, it fails after a timeout as Protractor reports Angular cannot be found on the page",
-    "//test2": "ng test --progress=false --watch=false && yarn e2e --configuration=ci && yarn e2e --configuration=ci-production",
-    "test": "ng test --progress=false --watch=false && yarn e2e --configuration=ci"
+    "test": "ng test --progress=false --watch=false && yarn e2e --configuration=ci && yarn e2e --configuration=ci-production"
```
:tada:

gkalpak commented on integration/cli-hello-world-ivy/e2e/src/app.e2e-spec.ts:
```diff
@@ -11,9 +15,4 @@ describe('cli-hello-world-ivy App', () => {
-  it('the percent pipe should work', () => {
-    page.navigateTo();
-    expect(page.getPipeContent()).toEqual('100 % awesome');
-  })
```
Why not add a pipe :wink:

filipesilva replied:
This integration test is meant to mimic https://github.com/angular/angular/tree/master/packages/core/test/bundling/hello_world and obtain similar sizes bundles. For that reason it doesn't have anything more than that app does.

I suppose this isn't clear because I have mentioned it exactly 0 times, and nothing in the test name reflects that either.

Maybe this test should be called `cli-minimal-ivy-hello-world`, WDYT? And reference that it's a version of that other one somewhere too.

gkalpak replied:
I see. It is definitely worth mentioning somewhere, because my general assumption is that `cli-*` projects in `integration/` are sort of equivalent to each other (which would make much more sense imo, especially given that their tracked payload sizes are included side-by-side in the same JSON file).

But if this one must much the one in `packages/`, your suggestions SGTM.

filipesilva replied:
Renamed to `cli-hello-world-ivy-minimal`, added a note to the integration test README, and to the app main.ts.

gkalpak commented on integration/cli-hello-world-ivy/src/app/app.component.ts:
```diff
@@ -2,9 +2,6 @@ import { Component } from '@angular/core';
   selector: 'app-root',
-  templateUrl: './app.component.html',
-  styleUrls: ['./app.component.css']
+  template: 'Welcome to cli-hello-world-ivy!',
```
Why such a minimal template?

filipesilva replied:
Same answer as the pipe comment above.

gkalpak commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -191,11 +191,8 @@ export class NgtscProgram implements api.Program {
-          `Listing specific routes is unsupported for now (got query for ${entryRoute})`);
-    }
     this.ensureAnalyzed();
+    // Listing specific routes is unsupported for now.
```
Maybe we could log a warning for now? Or is it safe to always return all lazy routes?

filipesilva replied:
I wouldn't like a warning because there's no way to turn it off.

I think it's ok to always return all lazy routes but will confirm with the team.

alxhub replied:
```suggestion
    // Listing specific routes is unsupported for now, so we erroneously
    // return all lazy routes instead (which should be okay for the CLI's
    // usage).
```

gkalpak reviewed (commented):
The permission changes (100755 --> 100644) in `integration/cli-hello-world-ivy/debug-test.sh` should be reverted (afaict).

filipesilva committed:
build: update render3 section of integration test readme

mary-poppins commented:
You can preview 6dbcf85 at https://pr28372-6dbcf85.ngbuilds.io/.

filipesilva commented:
@gkalpak the permissions change should be gone now.

mary-poppins commented:
You can preview cf3ecbc at https://pr28372-cf3ecbc.ngbuilds.io/.

mary-poppins commented:
You can preview e513ab5 at https://pr28372-e513ab5.ngbuilds.io/.

mary-poppins commented:
You can preview 778bea7 at https://pr28372-778bea7.ngbuilds.io/.

mary-poppins commented:
You can preview 5507426 at https://pr28372-5507426.ngbuilds.io/.

gkalpak commented on integration/cli-hello-world-ivy-minimal/README.md:
```diff
@@ -0,0 +1,27 @@
+# CliHelloWorldIvy
```
_CliHelloWorldIvy_ --> _CliHelloWorldIvyMinimal_ 😇

filipesilva replied:
Donesies 👍

gkalpak reviewed (approved):
👌

mary-poppins commented:
You can preview 61d6de9 at https://pr28372-61d6de9.ngbuilds.io/.

mary-poppins commented:
You can preview 4162517 at https://pr28372-4162517.ngbuilds.io/.

mary-poppins commented:
You can preview 1def325 at https://pr28372-1def325.ngbuilds.io/.

mary-poppins commented:
You can preview be6ef37 at https://pr28372-be6ef37.ngbuilds.io/.

mary-poppins commented:
You can preview 4b9641b at https://pr28372-4b9641b.ngbuilds.io/.
